### DISCUSSION TITLE
Slow rendering when displaying a lot of items

### DISCUSSION THREAD

**Discussion by denisgurev** [pocketbase/pocketbase#737] [Q&A] [answered]
I want to show 1000+ "cells". "cells" is my collection with 5 fields (posX, posY, name, description, user).

Unfortunately when I make a request to show the cells it takes 2-3s to render. Here is an excerpt from my code:

```
<script>
let cells = [];

async function cells() {
    cells = await client.records.getFullList("cells", 250);
}

cells();
</script>

{#each cells as cell (cell.id)}
   ...
{/each}
```

When I look in the inspect tool I see the api result really fast - 40-50ms. I also verified this with curl.

I'm still learning javascript so I'm not sure what I'm doing wrong. Is it possible the sdk to be slow or there is something wrong with my code?

_edit: I'm using sveltekit_

**Comment by ganigeorgiev** [comment, marked as answer]
It is hard to tell what takes so much time to render without additional context what you are doing in the each block.

I doubt that this is related to the JavaScript SDK, because the model resolver doesn't do anything "expensive" (see https://github.com/pocketbase/js-sdk/blob/master/src/models/Record.ts) and it shouldn't result in something like seconds difference.
You can verify it by making the GET request directly with `fetch()` and work with the raw json instead of the `Record` object.

My guess is that you have a lot of sub-components and/or conditional rendering in the `each` block and printing everything at once it is a little too much for svelte to process quickly.
There are several reported issues in the svelte repository that may be related, like https://github.com/sveltejs/svelte/pull/7663 or https://github.com/sveltejs/svelte/pull/3642.

While working on optimizing the listings in v0.8 I found out that "breaking" the updates in its own separate `task` helps significantly.

You could try changing your code to something like the following and see if there will be any improvement:

```html
<script>
let cells = [];

function breakTask() {
    return new Promise((resolve) => {
        setTimeout(resolve, 0);
    });
}

async function loadCells() {
    const items = await client.records.getFullList("cells", 250);
    
    cells = [];

    for (const item of items) {
        cells.push(item);
        cells = cells; // triggers svelte reactivity

        // break the update into a separate browser "task"
        await breakTask();
    }
}

loadCells();
</script>

{#each cells as cell (cell.id)}
   ...
{/each}
```

If the above doesn't work, you could try opening the browse's "Inspect tool > Performance" tab and record the function load call to check what exactly takes long "Scripting" times (usually marked with yellow/orange).

**Reply by denisgurev** [reply]
Wow, it is almost instant now! Yes I have a separate component for each field and also I'm using base64 data for the user avatar (it is a remote image fetched from a roblox url).

Thank you @ganigeorgiev!

Could you explain why this hack works. The promise doesn't do anything right? Is this a bug with svelte?

**Reply by ganigeorgiev** [reply]
It's not the `Promise` that instruct the browser to deffer/break the task. It's the `setTimeout` call.
For more details you could check this post from @malchata in https://web.dev/optimize-long-tasks/.
They use something similar but they've named the method `yieldToMain()` (and probably it is more correct).

>  Is this a bug with svelte?

I don't know. There are already several similar reported performance issues in the svelte repository, but I guess the svelte team are currently more focused on the sveltekit release.

**Reply by malchata** [reply]
Backing up @ganigeorgiev here: it's the `setTimeout` call that breaks up the task. However, there's considerable work being done with the [dedicated scheduler API](https://github.com/WICG/scheduling-apis/blob/main/explainers/prioritized-post-task.md), including [built-in yield and continuation](https://github.com/WICG/scheduling-apis/blob/main/explainers/yield-and-continuation.md). If (and when) this lands, It'd be an improvement over the `Promise`-based yielding function.
